src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,85 +1081,165 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
+                    // decompressed_size must equal to input sz before compress
+                    HS_REL_ASSERT_EQ(uint64_cast(mblk->hdr.h.src_context_sz),
+                                     uint64_cast(decompressed_size)); /* since decompressed_size is >=0 it
+                                                                         is safe to cast to uint64_t */
+                    HS_LOG(DEBUG, metablk,
+                           "[type={}] Successfully decompressed, compressed_sz: {}, src_context_sz: {}, "
+                           "decompressed_size: {}",
+                           mblk->hdr.h.type, uint64_cast(mblk->hdr.h.compressed_sz),
+                           uint64_cast(mblk->hdr.h.src_context_sz), decompressed_size);
                 }
 
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
-                       uint64_cast(mblk->hdr.h.context_sz));
+                cb(mblk, decompressed_buf, mblk->hdr.h.src_context_sz);
+            } else {
+                // There is use case that cb could be nullptr because client want to get its superblock via
+                // read api;
+                cb(mblk, buf, mblk->hdr.h.context_sz);
             }
-        } else {
-            HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
+
+            HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
+                   uint64_cast(mblk->hdr.h.context_sz));
+        }
+    } else {
+        HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
 #if 0
             // should never arrive here since we do assert on type before write to disk;
             HS_LOG_ASSERT( false, "[type={}] not registered for mblk found on disk. Skip this meta blk. ",
                       mblk->hdr.h.type);
 #endif
+    }
+}
+
+std::vector< meta_sub_type >
+MetaBlkService::topological_sort(std::unordered_map< meta_sub_type, std::vector< meta_sub_type > >& topological_graph) {
+    std::vector< meta_sub_type > result;
+    std::unordered_map< meta_sub_type, int > in_degree;
+    std::queue< meta_sub_type > q;
+
+    // Calculate in-degree of each vertex
+    for (const auto& [vertex, edges] : topological_graph) {
+        // we should make sure all the vertex in in_degree map;
+        // if vertex is not in the map, 0 will be assigned.
+        in_degree[vertex];
+        for (const auto& edge : edges) {
+            in_degree[edge]++;
         }
     }
 
-    if (do_comp_cb) {
-        // for each registered subsystem, do recovery complete callback;
-        for (auto& sub : m_sub_info) {
-            if (sub.second.comp_cb) {
-                sub.second.comp_cb(true);
-                HS_LOG(DEBUG, metablk, "[type={}] completion callback sent.", sub.first);
-            }
+    // Add vertices with in-degree 0 to the queue
+    for (const auto& [vertex, degree] : in_degree) {
+        if (degree == 0) { q.push(vertex); }
```

Comment:
the whole https://github.com/eBay/HomeStore/pull/200/files#diff-428b89aa263f097554416e42d736778169b33e2b6d72febef88b1c981b84bd0aR1086-R1102
can be moved here 
```
if (degree == 0) { 
    if (len(topological_graph[vertex]) ==0) 
          indep_subs.push_back(vertx)
     else 
           q.push(vertex); 
}
```

need to change the function signature as
```
bool MetaBlkService::topological_sort(std::unordered_map< meta_sub_type, std::vector< meta_sub_type > >& topological_graph,  std::vector< meta_sub_type> &indep_subs, std::vector< meta_sub_type > &ordered_subs)
```